packages/editor/src/hooks/pattern-partial-syncing.js

```diff
@@ -64,10 +65,45 @@ const withPartialSyncingControls = createHigherOrderComponent(
 	}
 );
 
+/**
+ * Adds an editOriginalPattern prop which allows the block to switch between post and pattern
+ * editing modes.
+ *
+ * @param {Component} BlockEdit Original component.
+ *
+ * @return {Component} Wrapped component.
+ */
+const withPatternOnlyRenderMode = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		if ( props.name !== 'core/block' ) {
+			return <BlockEdit { ...props } />;
+		}
+
+		const { onClick } = useLink( {
+			postId: props.attributes?.ref,
+			postType: 'wp_block',
+			canvas: 'edit',
+			editMode: 'focused',
+		} );
+
+		const newProps = {
+			...props,
+			editOriginalPattern: onClick,
+		};
+
+		return <BlockEdit { ...newProps } />;
+	}
+);
+
 if ( window.__experimentalPatternPartialSyncing ) {
 	addFilter(
 		'editor.BlockEdit',
 		'core/editor/with-partial-syncing-controls',
 		withPartialSyncingControls
 	);
+	addFilter(
+		'editor.BlockEdit',
+		'core/editor/with-pattern-edit-original-mode',
+		withPatternOnlyRenderMode
+	);
```

Comment:
Aside: this pattern of adding filters to blocks like that, is probably something we should avoid as much as we can if possible. It's probably something that creates performance issues.